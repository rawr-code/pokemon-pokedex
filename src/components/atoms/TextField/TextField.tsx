import { FC, useMemo, useState } from 'react'

// Atoms
import { Icons } from '@atoms'

// Utils
import { createTestIDs, cn } from '@utils'

const ownTestIDs = createTestIDs('TextField', ['root', 'input', 'action'])

// Types
interface TextFieldProps extends React.ComponentPropsWithoutRef<'input'> {
  className?: string
  classes?: {
    root?: string
  }
  testIDs?: { [k in keyof typeof ownTestIDs]?: string | undefined }
}

const TextField: FC<TextFieldProps> = ({
  type,
  placeholder,
  required,
  value,
  onChange,
  className = '',
  classes = {},
  testIDs = {},
}) => {
  const [showPass, setShowPass] = useState(false)

  const handleInputType = () => {
    setShowPass(p => !p)
  }

  const inputType = useMemo(() => {
    if (type === 'password') {
      if (!showPass) return 'password'
      return 'text'
    }
    return type
  }, [showPass, type])

  return (
    <div
      className={cn(
        'flex w-full rounded-xl border-2 border-gray-300',
        className,
        classes.root,
      )}
      data-testid={testIDs.root || ownTestIDs.root}
    >
      <input
        required={required}
        type={inputType}
        placeholder={placeholder}
        value={value}
        onChange={onChange}
        className="w-full rounded-xl bg-transparent px-4 py-[9.5px]"
        data-testid={testIDs.input || ownTestIDs.input}
      />
      {type === 'password' && (
        <button
          type="button"
          className="mr-3"
          onClick={handleInputType}
          data-testid={testIDs.action || ownTestIDs.action}
        >
          {showPass ? (
            <Icons.eyeClose size={24} color="stroke-pokedex-body" />
          ) : (
            <Icons.eye size={24} color="stroke-pokedex-body" />
          )}
        </button>
      )}
    </div>
  )
}

export default TextField
export { ownTestIDs as testIDs }
