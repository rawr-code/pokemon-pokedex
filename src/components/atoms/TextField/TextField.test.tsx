import { fireEvent, screen } from '@testing-library/react'
import { runCommonTests, render } from '@utils'

// Component
import TextField, { testIDs } from './TextField'

describe('Atom | TextField', () => {
  runCommonTests(TextField, { rootTestID: testIDs.root })

  it('should render the correct TextField', () => {
    expect.assertions(2)

    render(<TextField type="password" />)

    const elem = screen.getByTestId(testIDs.root)
    const input = screen.getByTestId<HTMLInputElement>(testIDs.input)

    fireEvent.change(input, { target: { value: '23' } })

    expect(input.value).toBe('23')
    expect(elem).toBeVisible()
  })

  it('show password', () => {
    expect.assertions(3)

    render(<TextField type="password" />)

    const elem = screen.getByTestId(testIDs.root)
    const input = screen.getByTestId<HTMLInputElement>(testIDs.input)
    const button = screen.getByTestId<HTMLButtonElement>(testIDs.action)

    expect(elem).toBeVisible()

    fireEvent.click(button)

    expect(input).toHaveAttribute('type', 'text')

    fireEvent.click(button)

    expect(input).toHaveAttribute('type', 'password')
  })
})
