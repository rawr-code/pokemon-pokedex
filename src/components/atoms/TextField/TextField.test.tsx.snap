// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Atom | TextField should match snapshot 1`] = `
<div>
  <div
    class="flex w-full rounded-xl border-2 border-gray-300 className classes.root"
    data-testid="TextField-root"
  >
    <input
      class="w-full rounded-xl bg-transparent px-4 py-[9.5px]"
      data-testid="TextField-input"
      value=""
    />
  </div>
</div>
`;
